test/udev-test.py

```diff
@@ -0,0 +1,2408 @@
+
+        if group:
+            try:
+                gid = grp.getgrnam(group).gr_gid
+            except KeyError:
+                gid = int(group)
+            assert gid == st.st_gid
+
+        if mode:
+            mode = int(mode, 8)
+            assert stat.S_IMODE(st.st_mode) == mode
+
+    def check_major_minor(self, st: os.stat_result) -> None:
+        if not self.exp_major_minor:
+            return
+        minor, major = (int(x) for x in self.exp_major_minor.split(':'))
+        assert st.st_rdev == os.makedev(minor, major)
+
+    def get_devnode(self) -> Path:
+        suffix = self.devnode if self.devnode else self.devpath.split('/')[-1]
+        return UDEV_DEV / suffix
+
+    def check_link_add(self, link: str, devnode: Path) -> None:
+        link = UDEV_DEV / link
+        tgt = link.parent / link.readlink()
+        assert devnode.samefile(tgt)
+
+    def check_link_nonexistent(self, link: str, devnode: Path) -> None:
+        link = UDEV_DEV / link
+
+        try:
+            tgt = link.parent / link.readlink()
+        except FileNotFoundError:
+            return
+
+        assert not devnode.samefile(tgt)
+
+    def check_add(self) -> None:
+        print(f'check_add {self.devpath}')
+
+        devnode = self.get_devnode()
+        st = devnode.stat(follow_symlinks=False)
+        assert stat.S_ISCHR(st.st_mode) or stat.S_ISBLK(st.st_mode)
+        self.check_permissions(st)
+        self.check_major_minor(st)
+
+        for link in self.exp_links or []:
+            self.check_link_add(link, devnode)
+
+        for link in self.not_exp_links or []:
+            self.check_link_nonexistent(link, devnode)
+
+    def check_link_remove(self, link: str) -> None:
+        link = UDEV_DEV / link
+        with pytest.raises(FileNotFoundError):
+            link.readlink()
+
+    def check_remove(self) -> None:
+        devnode = self.get_devnode()
+        assert not devnode.exists()
+
+        for link in self.exp_links or []:
+            self.check_link_remove(link)
+
+
+def listify(f):
+    def wrap(*args, **kwargs):
+        return list(f(*args, **kwargs))
+    return functools.update_wrapper(wrap, f)
+
+@listify
+def all_block_devs(exp_func) -> list[Device]:
+    # Create a device list with all block devices under /sys
+    # (except virtual devices and cd-roms)
+    # the optional argument exp_func returns expected and non-expected
+    # symlinks for the device.
+
+    for p in UDEV_SYS.glob('dev/block/*'):
+        tgt = os.readlink(p)
+        if re.search('/virtual/ | /sr[0-9]*$', tgt, re.VERBOSE):
+            continue
+
+        assert tgt.startswith('../../')
+        tgt = tgt[5:]
+
+        exp, not_exp = exp_func(tgt)
+        yield Device(devpath=tgt,
+                     exp_links=exp,
+                     not_exp_links=not_exp)
+
+
+@dataclasses.dataclass
+class Rules:
+    desc: str
+    devices: list[Device]
+    rules: str
+    repeat: int = 1
+    delay: Optional[float] = None
```

Comment:
float -> string or unsigned?

Reply:
There is no unsigned in python. I changed it to `int` now.